safeweb/http.go

```diff
@@ -0,0 +1,89 @@
+package safeweb
+
+import (
+	crand "crypto/rand"
+	"fmt"
+	"net"
+	"net/http"
+	"strings"
+
+	"github.com/gorilla/csrf"
+)
+
+type Options struct {
+	Hostname                    string
+	ApplicationHandler          http.Handler
+	PermittedCrossOriginHosts   []string
+	PermittedCrossOriginMethods []string
+
+	ServeHTTPS bool
+	// ServeHTTP                   bool
+	// RedirectHTTPS               bool
+
+	csrfProtect func(http.Handler) http.Handler
+}
+
+func (o *Options) setDefaults() {
+	if o.ApplicationHandler == nil {
+		o.ApplicationHandler = &http.ServeMux{}
+	}
+
+	if len(o.PermittedCrossOriginMethods) == 0 {
+		o.PermittedCrossOriginMethods = []string{"GET", "POST", "OPTIONS"}
+	}
+}
+
+type Server struct {
+	Options
+	h *http.Server
+}
+
+func wrapHandler(opts Options) func(http.Handler) http.Handler {
```

Comment:
let's also add any bits of https://github.com/tailscale/tailscale/blob/v1.62.0/tsweb/tsweb.go#L603 that are relevant

Reply:
Copied over the `X-Content-Type-Options` header as I understand the frame-options one is superseded by `Content-Security-Policy`'s `frame-*` toggles.